server/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -209,7 +211,7 @@ public T retrieve(String[] path, int index, Map<String, String> params, TrieMatc
             if (index >= path.length)
                 return null;
 
-            String token = path[index];
+            String token = RestUtils.decodePathComponent(path[index]);
```

Comment:
Unfortunately the path wasnt decoded prior to matching handlers. Ive got a few tests in RestControllerTests.java (below)

Reply:
This meant if the path was url encoded, it would not match handlers that it would if it was decoded. Tests exist below.